autoretrieve/autoretrieve.go

```diff
@@ -0,0 +1,322 @@
+package autoretrieve
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/application-research/estuary/util"
+	provider "github.com/filecoin-project/index-provider"
+	"github.com/filecoin-project/index-provider/metadata"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/multiformats/go-multihash"
+	"gorm.io/gorm"
+)
+
+type Autoretrieve struct {
+	gorm.Model
+
+	Handle            string `gorm:"unique"`
+	Token             string `gorm:"unique"`
+	LastConnection    time.Time
+	LastAdvertisement time.Time
+	PrivateKey        string `gorm:"unique"`
+	Addresses         string
+}
+
+type HeartbeatAutoretrieveResponse struct {
+	Handle            string         `json:"handle"`
+	LastConnection    time.Time      `json:"lastConnection"`
+	LastAdvertisement time.Time      `json:"lastAdvertisement"`
+	AddrInfo          *peer.AddrInfo `json:"addrInfo"`
+}
+
+type AutoretrieveListResponse struct {
+	Handle            string         `json:"handle"`
+	LastConnection    time.Time      `json:"lastConnection"`
+	LastAdvertisement time.Time      `json:"lastAdvertisement"`
+	AddrInfo          *peer.AddrInfo `json:"addrInfo"`
+}
+
+type AutoretrieveInitResponse struct {
+	Handle         string         `json:"handle"`
+	Token          string         `json:"token"`
+	LastConnection time.Time      `json:"lastConnection"`
+	AddrInfo       *peer.AddrInfo `json:"addrInfo"`
+}
+
+type SimpleEstuaryMhIterator struct {
+	offset int
+	Mh     []multihash.Multihash
+}
+
+func (m *SimpleEstuaryMhIterator) Next() (multihash.Multihash, error) {
+	if m.offset < len(m.Mh) {
+		hash := m.Mh[m.offset]
+		m.offset++
+		return hash, nil
+	}
+	return nil, io.EOF
+}
+
+// newIndexProvider creates a new index-provider engine to send announcements to storetheindex
+// this needs to keep running continuously because storetheindex
+// will come to fetch advertisements "when it feels like it"
+func NewAutoretrieveEngine(stopCh chan struct{}, tickInterval time.Duration, db *gorm.DB) (*AutoretrieveEngine, error) {
+	host, err := libp2p.New()
```

Comment:
this shouldnt be spinning up its own libp2p instance, it should just be using the host of the estuary node thats running it.

Reply:
Nice catch! It's leftover from when I was trying to spoof the autoretrieve server on advertisements. I'll remove this since this is not required.